src/cargo/core/registry.rs

```diff
@@ -174,14 +177,18 @@ impl<'cfg> PackageRegistry<'cfg> {
         sub_vec.push((id, deps));
     }
 
-    fn load(&mut self, source_id: &SourceId, kind: Kind) -> CargoResult<()> {
+    fn load(&mut self, source_id: &SourceId, kind: Kind, 
+            updated: &mut HashSet<String>) -> CargoResult<()> {
```

Comment:
Do you by any chance know where this machinery lives? I would be happy to reuse it.

Without this set, the registry is updated multiple times, if multiple packages are passed to `cargo update` ([test case](https://github.com/rust-lang/cargo/pull/1828/files#diff-fcf81cfe756ff54027fad304503760d2R833)), because the souces array passed to `add_sources` [here](https://github.com/rust-lang/cargo/pull/1828/files#diff-72d3ed8b640540c0f99dbe2877ef0ad6R84) can contain the same source (e.g. the registry) multiple times.